### PR TITLE
Admin dashboard

### PR DESCRIPTION
**Pull request by ajpulzone**
*I have already pulled the most current branch and resolved merge conflicts locally so hopefully that will make for a smooth transition.

Merge request:

1.) controllers/admin & routes
   i. set up namespace and routes for the admin dashboard, admin merchants and admin invoices show/index pages
   ii. set up folder system for controllers/admin/
   iii. created namespace controllers for admin invoices and admin merchants
   iv. renamed admin controller to dashboard controller to decrease confusion in regards to the namespace and the controller name. updated all associated files to be dashboard instead of admin.

2.) app/models/invoices
   i.) Created incomplete_invoices method and associated testing

3.) features/admin/index_spec
   i.) added testing for:
      a.) unique invoice ID's of all invoices that have items that have not shipped
      b.) each of those ID's is a link to that invoices admin show page
      c.) created at date formatting next to invoice ID's
      d.) used orderly to confirm the invoices were listed in descending order based on the created at date.
      d.) refactored "visit path" to use the updated namespace and advanced routing paths
      e.) refactored previous tests that had used html path verification to use advanced routing/namespace paths
      f.) refactored within blocks to use section identifiers of dashboard instead of admin
      
4.) views
   i.) Added blank index and show views for admin/merchants and admin/invoices
   ii.) added section for incomplete invoices
   iii.) refactored view to utilize namespace and advanced routing instead of hrefs.
   
Please let me know if you have any questions!

### CODE REVIEW

File: spec/models/invoice_spec.rb [hunk 1]
```diff
@@ -27,6 +27,41 @@
...
+        @invoice_item_4 = InvoiceItem.create!(item_id: @item_4.id, invoice_id: @invoice_2.id, quantity: 44567, unit_price: 45, status: 1)
+        @invoice_item_5 = InvoiceItem.create!(item_id: @item_5.id, invoice_id: @invoice_2.id, quantity: 1, unit_price: 10000000, status: 2)
+        @invoice_item_6 = InvoiceItem.create!(item_id: @item_6.id, invoice_id: @invoice_3.id, quantity: 738, unit_price: 90999, status: 2)
+      end
+
+      it "returns a list of all unique invoices that have items that have not been shipped from newest to oldest based on when
+        the invoice was created " do
+        expect(Invoice.incomplete_invoices).to eq([@invoice_1, @invoice_2])
+        expect(Invoice.incomplete_invoices).to_not eq([@invoice_2, @invoice_1])
+        expect(Invoice.incomplete_invoices).to_not include(@invoice_3)
```
Comment (Freeing3092) [inline spec/models/invoice_spec.rb]: The test for this method looks great, but I don't believe the expect statement on lines 62 and 63 are necessary as line 61 already tests the return value for the same object method call.

File: spec/features/admin/index_spec.rb [hunk 2]
```diff
@@ -55,33 +74,37 @@
...
+      expect(current_path).to eq(admin_invoices_path)
   end
 
   it "has a list of the top 5 customers who have conducted the largest number of successful
     transactions" do
-      visit "/admin"
+      visit "/admin/dashboard"
 
-      within("#admin-dashboard") do
+      within("#dashboard-customers") do
```
Comment (Freeing3092) [inline spec/features/admin/index_spec.rb:99]: It looks like the presence of the content expected in this `within` block is covered by the more robust `within` block on line 107. We might consider deleting this `within` block to DRY up our spec file.

File: spec/features/admin/index_spec.rb [hunk 3]
```diff
@@ -93,14 +116,57 @@
...
+      expect(current_path).to eq(admin_invoice_path("#{@invoice_2.id}"))
+  end
+
+  it "next to each invoice is the date created formatted like 'Monday, July 18, 2019
+    and the invoices are listed in order from oldest to newest" do
+    
+    visit "/admin/dashboard"
+
+      within("#dashboard-incomplete_invoices") do
+        expect("#{@invoice_2.id}").to appear_before("#{@invoice_1.id}")
```
Comment (Freeing3092) [inline spec/features/admin/index_spec.rb]: I don't believe this test is covering the correct order indicated in the User Story (Oldest to Newest). `@invoice_2` appears after `@invoice_1` in the `before :each` block; Meaning `@invoice_1` would be the oldest and accordingly should appear first.

This was not apparent looking at the `save_and_open_page` result as only the date is shown and not the time. But when I booted up the rails server with the csv data, I was able to confirm the page had the newest invoices first.

File: app/views/admin/dashboard/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    <% @customers.top_five_customers.each do |customer| %>
+  <li>
+   Name: <%= customer.first_name %> <%= customer.last_name %> | Successful Transactions: <%= customer.transaction_ct("success") %>
+  </li>
+<% end %>
+</ul>
+
+<ul id="dashboard-incomplete_invoices">
+  <p>Incomplete Invoices: </p>
+    <% @invoices.incomplete_invoices.reverse_order.each do |invoice| %>
```
Comment (Freeing3092) [inline app/views/admin/dashboard/index.html.erb]: It appears the `incomplete_invoices` method was already ordering the results by ascending order (this is the default). `reverse_order` changes the order to descending when we want to keep the order ascending per the User Story: "And I see that the list is ordered from oldest to newest".

See comment on line 164 of the `features/admin/index_spec.rb`

**Review by Freeing3092 — CHANGES_REQUESTED** [review]
Great job on these User Stories.

Two comments that I believe would require changes:
1. Sorting of Admin Dashbord Invoices (see comments in spec and view files)
2. Removal of coverage files.

File: app/views/admin/dashboard/index.html.erb (hunk 4 above)
Comment (ajpulzone) [reply app/views/admin/dashboard/index.html.erb]: Thank you! I got those mixed up so will fix locally and push up the changes.

File: spec/features/admin/index_spec.rb (hunk 2 above)
Comment (ajpulzone) [reply spec/features/admin/index_spec.rb:99]: I will combine the tests and delete the extra within block.

File: spec/features/admin/index_spec.rb (hunk 3 above)
Comment (ajpulzone) [reply spec/features/admin/index_spec.rb]: I will change the test to reflect the removal of .reverse_order in the view and push up the changes.

File: spec/models/invoice_spec.rb (hunk 1 above)
Comment (ajpulzone) [reply spec/models/invoice_spec.rb]: Will make changes and push up branch

**Review by Freeing3092 — APPROVED** [review]

**Pull request merged**